target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl839x.c

```diff
@@ -1817,6 +1817,84 @@ void rtl839x_set_receive_management_action(int port, rma_ctrl_t type, action_typ
 	}
 }
 
+static void rtl839x_led_init(struct rtl838x_switch_priv *priv)
+{
+	int i, pos;
+	u32 v, set;
+	u64 pm_copper = 0, pm_fiber = 0;
+	u32 setlen;
+	const __be32 *led_set;
+	char set_name[9];
+	struct device_node *node;
+	const char *mode;
+
+	node = of_find_compatible_node(NULL, NULL, "realtek,rtl8390-leds");
+	if (!node) {
+		pr_info("%s No compatible LED node found\n", __func__);
+		return;
+	}
+
+	for (i= 0; i < priv->cpu_port; i++) {
+		pos = (i << 1) % 32;
+		sw_w32_mask(0x3 << pos, 0, RTL839X_LED_FIB_SET_SEL_CTRL(i));
+		sw_w32_mask(0x3 << pos, 0, RTL839X_LED_COPR_SET_SEL_CTRL(i));
+
+		if (!priv->ports[i].phy)
+			continue;
+
+		if (priv->ports[i].phy_is_integrated)
+		pm_fiber |= BIT_ULL(i);
+			else
+		pm_copper |= BIT_ULL(i);
+
+		set = priv->ports[i].led_set;
+		sw_w32_mask(0, set << pos, RTL839X_LED_COPR_SET_SEL_CTRL(i));
+		sw_w32_mask(0, set << pos, RTL839X_LED_FIB_SET_SEL_CTRL(i));
+	}
+
+	sw_w32(0, RTL839X_LED_SET0_0_CTRL);
+	sw_w32(0, RTL839X_LED_SET0_0_CTRL - 4);
+	for (i = 0; i < 4; i++) {
+		pos = (i % 2)?15:0;
+		sprintf(set_name, "led_set%d", i);
+		led_set = of_get_property(node, set_name, &setlen);
+		if (!led_set || setlen != 12)
+			break;
+		v = be32_to_cpup(led_set) | be32_to_cpup(led_set + 1) << 5
+			| be32_to_cpup(led_set + 2)  << 10;
```

Comment:
Why do I need to configure all three port LED sets, if there could also be only one or two? If the user only provides one set configuration (because there is only one LED per port), this will silently fail.